### ISSUE TITLE
Configuration database does not accept connection strings

### ISSUE THREAD

**Issue by krisavi** [grafana/grafana#39422]
**What happened**:
Entered
`postgresql://grafana:grafanapass@host1:host1port,host2:host2port,host3:host3port/grafanadb?target_session_attrs=read-write`
into [database] url parameter

**What you expected to happen**:
It to detect I am using postgresql as it is correct schema for postgresql.
It gave error that "postgresql" is unknown database type, changed schema to "postgres", still unable to use multiple hosts. 
Expect it to connect to the RW node of the DB cluster.

**How to reproduce it (as minimally and precisely as possible)**:
Set up postgres cluster and add it by connection string.
```
[database]
url="postgresql://grafana:grafanapass@host1:host1port,host2:host2port,host3:host3port/grafanadb?target_session_attrs=read-write"
```
**Anything else we need to know?**:
https://www.postgresql.org/docs/13/libpq-connect.html Whole documentation about how to use connection string properly for postgesql.
Most of the frameworks even support it fully.

https://grafana.com/docs/grafana/latest/administration/set-up-for-high-availability/ 
In here you reference that the parameters have to be configured properly, but from database point of view, where I have set up DB cluster to be also HA, then Grafana configuration does not allow me to use cluster, but only one node. In case of master node change I would have to reconfigure grafana to rw node as well. That is bad approach and should be fixed on application side. 

**Environment**:
- Grafana version: 8.2.0~beta1
- Data source type & version: Postgres 13
- OS Grafana is installed on: Debian 11
- User OS & Browser: N/A
- Grafana plugins: N/A
- Others: N/A

**Comment by siepkes** [comment]
I've bumped in to the same issue.

The issue is that Grafana uses `pq` as it's PostgreSQL driver. The `pq` authors themselves state they can't provide reliable resolution of bugs anymore and recommend to use [pgx](https://github.com/jackc/pgx) instead. pgx does have multiple DB host support (see https://github.com/jackc/pgx/pull/545).

Even if one would provide Grafana with a PR to swap `pq` for `pgx` I doubt they would merge it. It would require extensive testing.

**Comment by krisavi** [comment]
I gave noticed yes that Grafana team prefers to use library that is in maintenance mode over doing some extensive testing when switching to some library that is getting updates and supporting newer versions of postgres. There could be a way to support both libraries and mark one as experimental, but so far reading #3676, I do not have high hopes for Grafana to ever support this functionality. Members of dev team refer to tools that do not fulfil this purpose, locking DB management to specific product or adding way too much complexity to the setup of connecting from service to RW node in DB cluster. There was also topic of it being difficult for grafana to determine RW node, but that is also handled by the connectionstring target_session_attrs attrbute, if other library were used.

I could not consider Grafana to be HA capable when it cannot handle HA DB clusters. What is the point of multiple Grafana nodes if one config DB node having issues can bring it all to halt.

I still consider it as a bug in connecting to DB and not a feature-request. Feature request would be to have separate connections for read-only and read write connectionstrings and having queries depending on their purpose to be load balanced across DB cluster. It is a bug if there is connectionstring possibility, but it does not accept what is being defined as connectionstring format and capabilities by DB developer.

**Comment by siepkes** [comment]
@krisavi I Agree that the developer in comment https://github.com/grafana/grafana/issues/3676#issuecomment-804296549 doesn't understand the issue. `proxysql`, `pgbounce`, etc. are all active components (which can break down and require configuration, management, monitoring, etc.) while a proper PostgreSQL driver such as `pgx` already has built-in support for client side failover. Which is a thousand times easier to use and operate. Also almost all PostgreSQL drivers in any language (JDBC, ODBC, etc.) have client side failover. 

One would think that the description of the libpq project: "For users that require new features **or reliable resolution of reported bugs**" would be enough to switch to `pgx` as DB driver. 

Kinda ironic that apparantly Grafana is [ISO 270001 certified](https://grafana.com/legal/security-compliance/) yet parts of their supply chain say: "Don't use us if you wan't reliable resolution of bugs"...

**Comment by siepkes** [comment]
I think the issue is still relevant, as outlined in comment https://github.com/grafana/grafana/issues/39422#issuecomment-1440471318.

**Comment by gabor** [comment]
hi @krisavi , sorry the `pgx` module is currently only used for the "postgres datasource plugin" functionality, the "postgres as grafana's database" functionality is still using the `lib/pq` module.

**Comment by krisavi** [comment]
> hi @krisavi , sorry the `pgx` module is currently only used for the "postgres datasource plugin" functionality, the "postgres as grafana's database" functionality is still using the `lib/pq` module.

Is there plan to get config DB to use pgx as well in future releases? Would guess that in config db side there could be less exceptions to handle. Don't see that there would be need for proxy support for it like it was created for datasource plugin.

**Comment by Pixis-Akshay-Gopani** [comment]
Hey @krisavi  have you tested for single host ?? For me its not working for even single host.

**Comment by dgarcia-collegeboard** [comment]
@gabor Hello,

Is there any update to this issue? I have compiled a list of related issues:

https://github.com/grafana/grafana/issues/13399
https://github.com/grafana/grafana/issues/89756
https://github.com/grafana/grafana/issues/3676

Earliest dating back to ~8 years ago. We have a need to configure grafana for HA setup, but there seems to be blockers as far as pointing grafana at Read-Only database instances. Outlined in those various issues, there is concern of various internal processes for grafana needing access to write to the DB. Has there been any work done to remove those dependencies on a writer DB so that grafana can successfully connect to a read-only DB? I have searched through documentation and not found answers that lead to a resolution.

It seems like it is not possible to point a grafana instance at a read-only db instance at this moment. For reference of my current setup:
**Grafana Version:** 9.4.3
**DB type**: AWS Aurora Postgres v16

I am willing to bump Grafana to a newer version if it means support for read-only is included 👍

I can successfully connect Grafana to Aurora's writer endpoint, but see failures when attempting to connect to the reader endpoint. I have attempted to use connection strings before stumbling upon this post and seeing that they are currently not supported. My current approach is similar to [#89756](https://github.com/grafana/grafana/issues/89756), however, in some of the issues, it seems that there seems to be support for read-only modes according to some commenters? Is the current solution to configure `grafana.ini` and set to read-only despite whatever user is logged in?

It seems that there is a lot of community interest in grafana supporting connection to read-only database interest, as issues like these seem to keep popping up throughout the years. It looks like your ping to the grafana backend team did not hotlink to any specific group of people, are they aware of this issue's existence? An update would be much appreciated. Thanks!

**Comment by krisavi** [comment]
> Hey [@krisavi](https://github.com/krisavi) have you tested for single host ?? For me its not working for even single host.

Single host has always worked for me.

**Comment by krisavi** [comment]
Yes this is still issue that should be solved by Grafana team.

**Comment by siepkes** [comment]
There is an open PR for this in `lib/pq`: https://github.com/lib/pq/pull/714 (in case switching to a maintained library like [pgx](https://github.com/jackc/pgx) is deemed too much work). But I doubt it will ever get merged since the last actual feature seems to be merged in 2023, there was only one (bugfix) commit in 2024 and no commits in 2025 so far.

On the plus side since `lib/pq` does not seem to be going anywhere applying that PR yourself seems easy enough to maintain since no (breaking) commits are added anyway.

**Comment by jisaitua** [comment]
This was added in lib/pq v1.11.0 — see [PR #1246](https://github.com/lib/pq/pull/1246) for details. Would it be possible to upgrade this dependency in Grafana?

**Comment by siepkes** [comment]
> This was added in lib/pq v1.11.0 — see [PR #1246](https://github.com/lib/pq/pull/1246) for details. Would it be possible to upgrade this dependency in Grafana?

I've patched a release into Grafana and am using it with client-side failover and it works great!

**Comment by macabu** [comment]
We've merged https://github.com/grafana/grafana/pull/125582 which updated `lib/pq` and should include the changes from https://github.com/lib/pq/pull/1246

It will be part of Grafana 13.1, and with backports for 13.0.x and 12.4.x (LTS)
